Update the Dashicon component to rely on the font that ships with WordPress

youknowriad opened the pull request:
closes #25230 

With the new @wordpress/icons package, the Dashicon component usage should be deprecated but even if it's deprecated, its code will have to remain on the project for Backward compatibility reasons. This is not a problem for any random component but it's a problem for the Dashicon because that component is huge (it includes all the SVGs) and it can't be tree-shaken for third-party usage (I know for example that Woo are struggling with this). 

In this PR, I update the implementation of the component to rely on the font and styles that already ship with WordPress. This has some advantages:

 - It ensures backward compatibility for third-party usage of this component in WordPress (as this is a hard rule for us). 
 - The bundle size is a lot smaller, so even if the component is indirectly used with npm package usage (like in the Button component), it doesn't harm the consumer's bundle size.

This is a breaking change though for the npm package itself because it means that component is basically useless outside of WordPress but breaking changes in npm are ok.

**Notes**

If you test the PR, you'll notice that some dashicons are not appearing, that's because the component here and the Dashicons version that ships with WordPress are not in sync. To solve this, we need to update WordPress Core Dashicons font to the last version with all the latest icons.

This also can't be merged now, we need to refactor all the mobile usage of the component and rely on the icons package instead. cc @Tug @hypest 

Thoughts?

youknowriad committed:
Update the Dashicon component to rely on the font that ships with WordPress

nerrad reviewed (commented):
Generally, this makes sense and helps keep the bundle-size down. However for back-compat usage consuming code must ensure any styles list `dashicons` as a dependency for the style registration. This likely is already taken care of in most cases in the wp-admin, but outside of wp-admin it won't be the case.

Along with the `CHANGELOG.md` mention of the breaking change, this might be worth a dev-note for whatever version of WP this lands in as well?

Also, is there any potential risk of validation errors for usage of `<DashIcon/>` that might be saved to post content?

aduth commented:
>This is a breaking change though for the npm package itself because it means that component is basically useless outside of WordPress but breaking changes in npm are ok.

So for people who _would_ want to use these icons outside of WordPress, is the suggested alternative that they use the `@wordpress/icons` package instead? Can that be included in the CHANGELOG?

> However for back-compat usage consuming code must ensure any styles list `dashicons` as a dependency for the style registration.

Shouldn't we be making `dashicons` a dependency of the `wp-components` stylesheet to address this? Or are we hoping to avoid that?

youknowriad commented:
> This likely is already taken care of in most cases in the wp-admin, but outside of wp-admin it won't be the case.

When are blocks going to be used outsiide fo wp-admin, I mean even if they have frontend styles, I don't expect blocks to use Dashicon component in the frontend or can they?

youknowriad commented:
> Shouldn't we be making dashicons a dependency of the wp-components stylesheet to address this? Or are we hoping to avoid that?

I was hoping I can avoid that because it's true that it's possible but probably very rare. that said, we can still do it.

aduth commented:
> > This likely is already taken care of in most cases in the wp-admin, but outside of wp-admin it won't be the case.
> 
> When are blocks going to be used outsiide fo wp-admin, I mean even if they have frontend styles, I don't expect blocks to use Dashicon component in the frontend or can they?

When we're saying "taken care of in wp-admin", are we talking about assuming the Dashicons font will already have been loaded by relying on it being some dependency of another stylesheet?

Tug commented:
Thanks for bringing us in the loop!

> This also can't be merged now, we need to refactor all the mobile usage of the component and rely on the icons package instead.

Ok 👍Should we create the branch, or do you plan to tackle it?

youknowriad commented:
@Tug I think it could be done in small PRs refactoring the existing usage step by step. Feel free to tackle some of these.

Example PRs:

#19838 
#19834 
#19808

nerrad commented:
> When are blocks going to be used outsiide fo wp-admin, I mean even if they have frontend styles, I don't expect blocks to use Dashicon component in the frontend or can they

For me it's less about blocks being used more outside the admin and more about the Dashicons _component_ being used outside the admin (as a part of a frontend client side rendering). We could treat this as similar to "package" use but if `dashicons` style is already a dependency on the `wp-components` style then at least in the WordPress context there should be no apparent breakage. 

> When we're saying "taken care of in wp-admin", are we talking about assuming the Dashicons font will already have been loaded by relying on it being some dependency of another stylesheet?

Ya, it's a dependency of a number of styles loaded in the wp-admin context so I'm not aware of (without completely verifying explicitly) of any `wp-admin` route that won't have the `dashicons` handle enqueued. With that said, like any dependencies, I don't think we should _rely_ on that _always_ being the case.

Even with my comments here, I'd still like to see this work done. My comments are just more of awareness of potential impact. So to summarize, should we go ahead with this change, these are the things from my perspective I think we'll want to cover:

- It's a breaking change for the package (so should be communicated as such).
- Make the WordPress core `dashicons` handle a dependency of `wp-components`.
- Publish a devnote about the change to surface for developers that _might_ have used the `<DashIcons/>` component in some custom react app _in_ the WordPress environment but outside of the wp-admin context.

youknowriad commented:
> If you test the PR, you'll notice that some dashicons are not appearing, that's because the component here and the Dashicons version that ships with WordPress are not in sync. To solve this, we need to update WordPress Core Dashicons font to the last version with all the latest icons.

@ntwb Do you think you can help with this? Update the dashicon font in Core with all the icons including the ones created for Gutenberg?

jasmussen commented:
> Update the dashicon font in Core with all the icons including the ones created for Gutenberg?

It _should_ be "enough" to move all the SVGs from the `sources/svg/gutenberg` into the `sources/svg` folder. Alternately the build script can be made to include the gutenberg subfolder when compiling the icon font, which it currently ignores.

youknowriad committed:
Remove unused stylesheet import

jasmussen commented:
I created https://github.com/WordPress/dashicons/pull/402 to graduate the Gutenberg icons.

jasmussen commented:
@youknowriad https://github.com/WordPress/dashicons/pull/402 has been merged. What are next steps here?

youknowriad commented:
The next steps is probably a patch for Core with the updated icon font.

sgomes commented:
Hey @youknowriad! Thank you for working on this, I think it's going to be a nice perf improvement for Gutenberg! 👍 

Looking at the source, it appears this will **require** the dashicons font to be loaded for icons to display correctly, correct? That should be a reasonable assumption for most Gutenberg usage, but for things like Gutenboarding that font isn't currently being loaded. In fact, the fact that this is a concern at all betrays what seems like a bit of a violation of abstraction of the package model, as there's now an implicit dependency that needs to be loaded through other means.

For Gutenboarding, the obvious solution would be to just load the font, but I was personally hoping to avoid that, as [icon fonts have a number of accessibility and rendering issues](https://cloudfour.com/thinks/seriously-dont-use-icon-fonts/) (old article, but still quite accurate).

What do you think is the best approach here? Should Gutenboarding find a way of hot-swapping `components/dashicons` with an alternative implementation that e.g. uses SVG sprites? Should we split out `dashicons` into its own `@wordpress/dashicons` component, with multiple implementations, or perhaps even multiple packages (e.g. `@wordpress/dashicons-iconfont` and `@wordpress/dashicons-svgsprite`) that could be hot-swapped via `webpack` config?

cc @simison @sirreal @Automattic/luna

youknowriad commented:
> Looking at the source, it appears this will require the dashicons font (and associated CSS) to be loaded for icons to display correctly, correct?

Yes, that requirement is not something we can remove at the moment because it's part of  the block registration API (and other icon APIs) even if we remove all the usage from the Gutenberg repository itself.

If for your own explorations you don't use the dashicons at all or don't use third party blocks that rely on dashicons (which is very hard to ensure), you could consider removing that dependency. I'd recommend to not do it though because it's very easy for third-party code to rely on these.

youknowriad commented:
> Swapping

One solution could be to copy the component removed in this PR and replace the "imports" (potentially possible using webpack configs)

senadir commented:
I'm having a hard time debugging why is `@wordpress/components@9.5.0` still loading dashicons, as far as I remember it was fully migrated off that package right? is it being imported somewhere in the code?

senadir commented:
I can see it's still being imported in `@wordpress/components/icon` to account for it, will have to find a way in webpack to remove it.

youknowriad committed:
Add changelog and dependency

youknowriad committed:
Fix unit tests

youknowriad committed:
Fix end2end tests and linting

youknowriad commented:
Now that the Dashicon font is updated on Core and frozen, I think it's time to revisit this PR. It seems ready for me.

hypest commented:
> Now that the Dashicon font is updated on Core and frozen, I think it's time to revisit this PR. It seems ready for me.

👋 @youknowriad , I wonder if you could add some testing instructions to help verify what icons are expected to work off the font?

@Tug or @ceyhun , can you have a new look on this PR? I haven't looked deeper but, I wonder if the mobile app will also need to embed the font that holds the icons and break if the font is not there?

youknowriad commented:
> I wonder if you could add some testing instructions to help verify what icons are expected to work off the font?

Do you have any ideas on mind. What would this test look like?

> can you have a new look on this PR? I haven't looked deeper but, I wonder if the mobile app will also need to embed the font that holds the icons and break if the font is not there?

At this point, I expect that our usage of Dashicons on Gutenberg Core to be pretty small. It's still not entirely replaced with SVGs but if you find things that are still relying on Dashicons we can replace them with SVGs to make it work cross platform.

ceyhun commented:
> Do you have any ideas on mind. What would this test look like?

I'm not sure if this is the optimal way but what I did was searching for `<Dashicon` and one of the places was here: https://github.com/WordPress/gutenberg/blob/8053ebc0b61f6ab2820e2de9b9ea7f10cddb0ff9/packages/components/src/icon/index.js#L21-L29

Than I searched for `<Icon` in `.native.js` files and found a string prop, this crashed the app when upload of an image failed for `Gallery` block:

https://github.com/WordPress/gutenberg/blob/7532a485b400f86638145b71f94f6f717e5add25/packages/block-library/src/gallery/gallery-image.native.js#L230-L233


There were also other places where

youknowriad committed:
Replace more dashicons

youknowriad commented:
@ceyhun I replaced these with svg usage, I also replaced a few other places. Let me know if you find any place I missed where dashicons are still being used.

I also noticed that on master, folks are continuing to use Dashicons for new code, so I'd like to move with this forward in order to officially deprecate the component and encourage folks to use the SVGs instead.

youknowriad commented:
Anyone to approve this :)

jasmussen reviewed (approved):
This tests well for me, let's get this in.

ceyhun commented:
> @ceyhun I replaced these with svg usage, I also replaced a few other places. Let me know if you find any place I missed where dashicons are still being used.

Thanks for finding and fixing the other places @youknowriad! There still could be other places we missed but it looks good for now.

tellthemachines commented on packages/components/CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@
 ### Breaking Change
 
 - `NumberControl` no longer automatically transforms values when rendering `value` into a `<input />` HTML element.
+- `Dashicon` component no longer renders SVGs. If you rely on this component, make sure to load the dashicon font.
```
Shouldn't this have gone under `Unreleased`?

youknowriad replied:
yep, mistake 😬

tellthemachines replied:
No worries, I bumped the major version in last week's release 😄
